### PR TITLE
Add Semux Blockchain

### PR DESCRIPTION
**Pull request by cryptokat**
* Add Semux Address, Signer, Transaction, Protos
* Add Semux unit tests

## Blockchain checklist

* [x] Add the coin definition to `coins.json` and `TWCoinType`.
* [x] Implement functionality in C++. Put it in a subfolder of `src/`.
  * [x] Address \(if necessary\)
  * [x] Transaction \(if necessary\)
  * [x] Signer
* [x] Write unit tests. Put them in a subfolder of `tests/`.
* [x] Implement address validation and derivation in `src/Coin.cpp`.
* [x] Write interface header in `include/TrustWalletCore` and implement the interface in `src/interface`.
  * [x] Address interface \(if necessary\).
  * [x] Signing interface.
* [x] Validate generated code in Android an iOS projects. Write integration tests for each.
* [x] Add a coin icon to "Supported Blockchains" section in `README.md`
  * [x] Upload coin icon to [TrustWallet/tokens](https://github.com/TrustWallet/tokens) if necessary

### CODE REVIEW

File: src/Semux/Signer.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#include "../Hash.h"
+#include "../PrivateKey.h"
+
+namespace TW::Semux {
+
+/// Helper class that performs Semux transaction signing.
+class Signer {
+  public:
+    /// Signs the given transaction.
+    void sign(const PrivateKey &privateKey, Transaction &transaction) const noexcept;
```
Comment (hewigovens) [inline src/Semux/Signer.h]: let's mark this as `static`

File: src/Semux/Transaction.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    std::vector<uint8_t> getPreImage() const;
+    std::vector<uint8_t> getHash() const;
+
+  private:
+    void writeBytes(const Data bytes, Data &data) const {
+        encodeVarInt(bytes.size(), data);
+        data.insert(data.end(), bytes.begin(), bytes.end());
+    }
+
+    void writeLong(const uint64_t l, Data &data) const {
```
Comment (hewigovens) [inline src/Semux/Transaction.h]: can we reuse `encode64BE` and `encode32BE` in `src/BinaryCoding.h`?

**Comment by hewigovens** [comment]
looks good overall, can you add test case to `CoinAddressDerivationTests.[swift|kt|test.ts]` to fix the CI? It's also nice to:

1. Add a Swift signing test case
2. cross verify mnemonic -> address with other wallets (e.g.: ledger / trezor)

**Comment by creamyavocado** [comment]
@hewigovens when it will be available in the wallet?

**Comment by victorx716** [comment]
> @hewigovens when it will be available in the wallet?

Any update on this?

**Pull request merged**